src/hotspot/cpu/ppc/stubGenerator_ppc.cpp

```diff
@@ -3544,6 +3544,462 @@ class StubGenerator: public StubCodeGenerator {
+
+    // The upper 32 bits of the non-pointer parameter registers are not
+    // guaranteed to be zero, so mask off those upper bits.
+    __ clrldi(sp, sp, 32);
+    __ clrldi(sl, sl, 32);
+    __ clrldi(dp, dp, 32);
+    __ clrldi(isURL, isURL, 32);
+
+    // Load constant vec registers that need to be loaded from memory
+    __ load_const_optimized(const_ptr, (address)&bitposLUT_val, tmp_reg);
+    __ lxv(bitposLUT, 0, const_ptr);
+    if (PowerArchitecturePPC64 >= 10) {
+        __ load_const_optimized(const_ptr, (address)&p10_pack_permute_val, tmp_reg);
+    } else {
+        __ load_const_optimized(const_ptr, (address)&pack_rshift_val, tmp_reg);
+        __ lxv(pack_rshift->to_vsr(), 0, const_ptr);
+        __ load_const_optimized(const_ptr, (address)&pack_lshift_val, tmp_reg);
+        __ lxv(pack_lshift->to_vsr(), 0, const_ptr);
+        __ load_const_optimized(const_ptr, (address)&pack_permute_val, tmp_reg);
+    }
+    __ lxv(pack_permute, 0, const_ptr);
+
+    // Splat the constants that can use xxspltib
+    __ xxspltib(vec_0s->to_vsr(), 0);
+    __ xxspltib(vec_4s->to_vsr(), 4);
+    __ xxspltib(vec_8s->to_vsr(), 8);
+    __ xxspltib(vec_0xfs, 0xf);
+    if (PowerArchitecturePPC64 >= 10) {
+        __ xxspltib(vec_0x3fs->to_vsr(), 0x3f);
+    }
+
+    // The rest of the constants use different values depending on the
+    // setting of isURL
+    __ cmpwi(CCR0, isURL, 0);
+    __ beq(CCR0, not_URL);
+
+    // isURL != 0 (true)
+    __ load_const_optimized(const_ptr, (address)&offsetLUT_URL_val, tmp_reg);
+    __ lxv(offsetLUT, 0, const_ptr);
+    __ load_const_optimized(const_ptr, (address)&maskLUT_URL_val, tmp_reg);
+    __ lxv(maskLUT, 0, const_ptr);
+    __ xxspltib(vec_special_case_char->to_vsr(), '_');
+    __ xxspltib(vec_special_case_offset, (unsigned char)US);
+    __ b(calculate_size);
+
+    // isURL = 0 (false)
+    __ bind(not_URL);
+    __ load_const_optimized(const_ptr, (address)&offsetLUT_val, tmp_reg);
+    __ lxv(offsetLUT, 0, const_ptr);
+    __ load_const_optimized(const_ptr, (address)&maskLUT_val, tmp_reg);
+    __ lxv(maskLUT, 0, const_ptr);
+    __ xxspltib(vec_special_case_char->to_vsr(), '/');
+    __ xxspltib(vec_special_case_offset, (unsigned char)SLS);
+
+    __ bind(calculate_size);
+
+    // Don't handle the last 4 characters of the source, because this
+    // VSX-based algorithm doesn't handle padding characters.  Also the
+    // vector code will always write 16 bytes of decoded data on each pass,
+    // but only the first 12 of those 16 bytes are valid data (16 base64
+    // characters become 12 bytes of binary data), so for this reason we
+    // need to subtract an additional 8 bytes from the source length, in
+    // order not to write past the end of the destination buffer.  The
+    // result of this subtraction implies that a Java function in the
+    // Base64 class will be used to process the last 12 characters.
+    __ sub(sl, sl, sp);
+    __ subi(sl, sl, 12);
```

Comment:
I think we should subtract 4, now. srawi will round it down below. We have no guarantee that we can subract more than 4 without getting negative value.

Reply:
In the original paper this code is based upon, they subtract 12 because of the overwrite issue.  This is discussed in the preceding code comment as well.  So I think that needs to be retained, but I do need to check for a negative after the subtract.  

According to the `srawi.` specification:

>  CA and CA32 are set to 1 if the low-order 32 bits of (RS) contain a negative number and any 1-bits are shifted out of position 63; otherwise CA and CA32 are set to 0.

Because the `sub` instruction is a 64-bit subtract, all of the upper bits should be 1's if sl is negative after the subtract, so I think the `srawi.` should catch the negative case if I also check CA after the srawi, via:
```
   __ srawi_(sl, sl, block_size_shift);
   // if XER CA is set, sl was less than zero.
   __ mcrxrx(CCR2); // moves XER's OV, OV32, CA, CA32 to CCR2's LT, GT, EQ, SO bits, respectively.
   __ beq_predict_not_taken(CCR2, unrolled_loop_exit);
   __ beq_predict_not_taken(CCR0, unrolled_loop_exit);
```